# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [3]:
import pandas as pd

import seaborn as sns
from seaborn import load_dataset

import patsy
import matplotlib.pyplot as plt
import numpy as np

import statsmodels.formula.api as smf
import statsmodels.api as sm

In [4]:
df = pd.read_csv('previsao_de_renda.csv')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


#### Aqui foi criado a regressão com base nos valores separados e modificados no dmatrices, porém anteriormente foi tentato utilizar o smf.ols na geração da regressão e obtive erro dizendo que a quantidade de linhas de y e X estava diferente, não consegui tratar o erro diretamente e mudei para o sm.OLS

In [6]:
# Separando os valores com o dmatrices do patsy e colocando as variáveis qualitativas em dummies
y, X = patsy.dmatrices('np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel) + qtd_filhos + \
            C(tipo_renda) + C(educacao) + C(estado_civil) + C(tipo_residencia) + idade + tempo_emprego + \
            qt_pessoas_residencia', df)

# Criando a regressão com base no dmatrices
reg_log = sm.OLS(y, X).fit()
reg_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     287.5
Date:                Mon, 14 Aug 2023   Prob (F-statistic):               0.00
Time:                        10:04:26   Log-Likelihood:                -13568.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12402   BIC:                         2.737e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               6.5847      0.235     28.006      0.000       6.124       7.046
C(sexo)[T.M]                            0.7874      0.015     53.723      0.000       0.759       0.816
C(posse_de_veiculo)[T.True]             0.0441      0.014      3.119      0.002       0.016       0.072
C(posse_de_imovel)[T.True]              0.0829      0.014      5.926      0.000       0.055       0.110
C(tipo_renda)[T.Bolsista]               0.2209      0.241      0.916      0.360      -0.252       0.694
C(tipo_renda)[T.Empresário]             0.1551      0.015     10.387      0.000       0.126       0.184
C(tipo_renda)[T.Pensionista]           -0.3087      0.241     -1.280      0.201      -0.782       0.164
C(tipo_renda)[T.Servidor público]       0.0576      0.022      2.591      0.010       0.014       0.101
C(educacao)[T.Pós graduação]            0.1071      0.159      0.673      0.501      -0.205       0.419
C(educacao)[T.Secundário]              -0.0141      0.072     -0.196      0.844      -0.155       0.127
C(educacao)[T.Superior completo]        0.0938      0.072      1.298      0.194      -0.048       0.235
C(educacao)[T.Superior incompleto]     -0.0432      0.078     -0.555      0.579      -0.196       0.109
C(estado_civil)[T.Separado]             0.3265      0.111      2.935      0.003       0.108       0.545
C(estado_civil)[T.Solteiro]             0.2672      0.109      2.453      0.014       0.054       0.481
C(estado_civil)[T.União]               -0.0343      0.025     -1.361      0.174      -0.084       0.015
C(estado_civil)[T.Viúvo]                0.3689      0.116      3.191      0.001       0.142       0.596
C(tipo_residencia)[T.Casa]             -0.0442      0.054     -0.815      0.415      -0.150       0.062
C(tipo_residencia)[T.Com os pais]      -0.0258      0.060     -0.427      0.670      -0.144       0.093
C(tipo_residencia)[T.Comunitário]      -0.1196      0.105     -1.135      0.256      -0.326       0.087
C(tipo_residencia)[T.Estúdio]           0.0665      0.099      0.670      0.503      -0.128       0.261
C(tipo_residencia)[T.Governamental]    -0.0571      0.066     -0.865      0.387      -0.186       0.072
qtd_filhos                             -0.2658      0.108     -2.463      0.014      -0.477      -0.054
idade                                   0.0053      0.001      6.624      0.000       0.004       0.007
tempo_emprego                           0.0617      0.001     59.398      0.000       0.060       0.064
qt_pessoas_residencia                   0.2926      0.108      2.716      0.007       0.081       0.504
==============================================================================
Omnibus:                        0.858   Durbin-Watson:                   2.023
Prob(Omnibus)

#### Para algumas modificações os valores faltantes sempre atrapalham um pouco, então será substituido os valores faltantes pela média dos valores. Em um projeto mais aprofundado isso pode ser revisto.

In [7]:
# Verificando valores faltantes no DataFrame
df.isna().sum()

Unnamed: 0                  0
data_ref                    0
id_cliente                  0
sexo                        0
posse_de_veiculo            0
posse_de_imovel             0
qtd_filhos                  0
tipo_renda                  0
educacao                    0
estado_civil                0
tipo_residencia             0
idade                       0
tempo_emprego            2573
qt_pessoas_residencia       0
renda                       0
dtype: int64

In [8]:
# Capturando a média dos valores com o mean
tempo_emprego_medio = df['tempo_emprego'].mean()
print(tempo_emprego_medio)

# Preenchendo os valores faltantes com o fillna passando o tempo_emprego_medio como parâmetro
df['tempo_emprego'] = df['tempo_emprego'].fillna(tempo_emprego_medio)
df.isna().sum() # Verificando se realmente não tem mais valores faltantes

7.722634652121815


Unnamed: 0               0
data_ref                 0
id_cliente               0
sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
qt_pessoas_residencia    0
renda                    0
dtype: int64

In [9]:
# Verificando os valores mais frequentes com uso do mode (moda de minhas variáveis em um DF)
modas = df.mode()
modas.head(1)

Unnamed: 0    data_ref  id_cliente sexo posse_de_veiculo posse_de_imovel   
0           0  2015-01-01      5573.0    F            False            True  \

   qtd_filhos   tipo_renda    educacao estado_civil tipo_residencia  idade   
0         0.0  Assalariado  Secundário       Casado            Casa   40.0  \

   tempo_emprego  qt_pessoas_residencia   renda  
0       7.722635                    2.0  728.96

#### As variáveis sexo, posse_de_veiculo, tipo_renda, estado_civil NÃO foram alteradas, pois já são a moda

#### As variáveis posse_de_imovel, educacao e tipo_residencia FORAM alteradas para ser a casela de referência

In [10]:
# Dividindo novamente as variáveis no dmatrices agora passando o Treatment para nova casela de referência
y, X = patsy.dmatrices("np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel, Treatment(1)) + \
            qtd_filhos + C(tipo_renda) + C(educacao, Treatment('Secundário')) + \
            C(estado_civil) + C(tipo_residencia, Treatment('Casa')) + idade + tempo_emprego + \
            qt_pessoas_residencia", df)

# Criando regressão com base no dmatrices
reg_log = sm.OLS(y, X).fit()
reg_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.348
Model:                            OLS   Adj. R-squared:                  0.347
Method:                 Least Squares   F-statistic:                     333.6
Date:                Mon, 14 Aug 2023   Prob (F-statistic):               0.00
Time:                        10:04:27   Log-Likelihood:                -16212.
No. Observations:               15000   AIC:                         3.247e+04
Df Residuals:                   14975   BIC:                         3.266e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       6.6821      0.212     31.586      0.000       6.267       7.097
C(sexo)[T.M]                                                    0.7947      0.014     57.710      0.000       0.768       0.822
C(posse_de_veiculo)[T.True]                                     0.0351      0.013      2.693      0.007       0.010       0.061
C(posse_de_imovel, Treatment(1))[T.False]                      -0.0897      0.013     -6.978      0.000      -0.115      -0.064
C(tipo_renda)[T.Bolsista]                                       0.2075      0.238      0.870      0.384      -0.260       0.675
C(tipo_renda)[T.Empresário]                                     0.1543      0.015     10.457      0.000       0.125       0.183
C(tipo_renda)[T.Pensionista]                                   -0.1910      0.021     -9.123      0.000      -0.232      -0.150
C(tipo_renda)[T.Servidor público]                               0.0571      0.022      2.602      0.009       0.014       0.100
C(educacao, Treatment('Secundário'))[T.Primário]                0.0719      0.056      1.278      0.201      -0.038       0.182
C(educacao, Treatment('Secundário'))[T.Pós graduação]           0.1268      0.141      0.902      0.367      -0.149       0.402
C(educacao, Treatment('Secundário'))[T.Superior completo]       0.1198      0.013      9.455      0.000       0.095       0.145
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    -0.0305      0.031     -0.978      0.328      -0.092       0.031
C(estado_civil)[T.Separado]                                     0.2754      0.107      2.568      0.010       0.065       0.486
C(estado_civil)[T.Solteiro]                                     0.2294      0.105      2.181      0.029       0.023       0.436
C(estado_civil)[T.União]                                       -0.0382      0.023     -1.662      0.097      -0.083       0.007
C(estado_civil)[T.Viúvo]                                        0.2855      0.108      2.637      0.008       0.073       0.498
C(tipo_residencia, Treatment('Casa'))[T.Aluguel]                0.0312      0.052      0.600      0.548      -0.071       0.133
C(tipo_residencia, Treatment('Casa'))[T.Com os pais]            0.0196      0.029      0.676      0.499      -0.037       0.076
C(tipo_residencia, Treatment('Casa'))[T.Comunitário]           -0.0795      0.090     -0.884      0.377      -0.256       0.097
C(tipo_residencia, Treatment('Casa'))[T.Estúdio]                0.0931      0.079      1.181      0.238      -0.061       0.248
C(tipo_residencia, Treatment('Casa'))[T.Governamental]          0.0299      0.034      0.866      0.386      -0.038       0.097
qtd_filhos 

#### Todas paracem fazer sentido, incluso a posse_de_imovel que está como false, isso quer dizer que as pessoas que não tem imovel vai ter um valor negativo se comparado com as que tem.

2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  

    

In [11]:
# Removendo a variável tipo_residencia, que é a menos significante
y, X = patsy.dmatrices("np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel, Treatment(1)) + \
            qtd_filhos + C(tipo_renda) + C(educacao, Treatment('Secundário')) + \
            C(estado_civil) + idade + tempo_emprego + \
            qt_pessoas_residencia", df)

# Criando regressão com base no dmatrices sem o tipo_residencia
reg_log = sm.OLS(y, X).fit()
reg_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.348
Model:                            OLS   Adj. R-squared:                  0.347
Method:                 Least Squares   F-statistic:                     421.3
Date:                Mon, 14 Aug 2023   Prob (F-statistic):               0.00
Time:                        10:04:27   Log-Likelihood:                -16214.
No. Observations:               15000   AIC:                         3.247e+04
Df Residuals:                   14980   BIC:                         3.262e+04
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       6.6829      0.211     31.603      0.000       6.268       7.097
C(sexo)[T.M]                                                    0.7958      0.014     57.955      0.000       0.769       0.823
C(posse_de_veiculo)[T.True]                                     0.0350      0.013      2.686      0.007       0.009       0.061
C(posse_de_imovel, Treatment(1))[T.False]                      -0.0870      0.013     -6.886      0.000      -0.112      -0.062
C(tipo_renda)[T.Bolsista]                                       0.2075      0.238      0.870      0.384      -0.260       0.675
C(tipo_renda)[T.Empresário]                                     0.1547      0.015     10.502      0.000       0.126       0.184
C(tipo_renda)[T.Pensionista]                                   -0.1902      0.021     -9.093      0.000      -0.231      -0.149
C(tipo_renda)[T.Servidor público]                               0.0580      0.022      2.645      0.008       0.015       0.101
C(educacao, Treatment('Secundário'))[T.Primário]                0.0696      0.056      1.239      0.215      -0.041       0.180
C(educacao, Treatment('Secundário'))[T.Pós graduação]           0.1250      0.140      0.889      0.374      -0.150       0.400
C(educacao, Treatment('Secundário'))[T.Superior completo]       0.1192      0.013      9.421      0.000       0.094       0.144
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    -0.0309      0.031     -0.992      0.321      -0.092       0.030
C(estado_civil)[T.Separado]                                     0.2771      0.107      2.584      0.010       0.067       0.487
C(estado_civil)[T.Solteiro]                                     0.2317      0.105      2.203      0.028       0.026       0.438
C(estado_civil)[T.União]                                       -0.0380      0.023     -1.655      0.098      -0.083       0.007
C(estado_civil)[T.Viúvo]                                        0.2862      0.108      2.645      0.008       0.074       0.498
qtd_filhos                                                     -0.2378      0.104     -2.277      0.023      -0.443      -0.033
idade                                                           0.0051      0.001      6.755      0.000       0.004       0.007
tempo_emprego                                                   0.0618      0.001     60.529      0.000       0.060       0.064
qt_pessoas_residencia                                           0.2618      0.104      2.511      0.012       0.057       0.466
==============================================================================
Omnibus:                        0.704   Durbin-Watson:      

#### Não alterou muito, o R² está o mesmo e estranhamente o R² Ajustado também, mesmo com a remoção de uma variável. Mas isso pode ser um bom sinal e mostra que a variável removida não fazia grande diferença na explicação dos resultados.

3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 

In [12]:
# Criando um filtro do pandas para ajudar a verificar quais valores precisam ser removidos
pvalues_series = pd.Series(reg_log.pvalues, index=reg_log.model.exog_names)
pvalues_filtrado = pvalues_series[pvalues_series >= 0.05]
pvalues_filtrado

C(tipo_renda)[T.Bolsista]                                      0.384207
C(educacao, Treatment('Secundário'))[T.Primário]               0.215327
C(educacao, Treatment('Secundário'))[T.Pós graduação]          0.373756
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    0.321274
C(estado_civil)[T.União]                                       0.097945
dtype: float64

In [13]:
# Criando variável pra guardar os valores que quero remover do meu DF
ex_educacao = ['Primário', 'Pós graduação', 'Superior incompleto']
ex_tipo_renda = ['Bolsista']
ex_estado_civil = ['União']

''' Aqui está sendo usado a máscara booleana, que inverte o resultado passado, ou seja, 
está invertendo o resultado de isin, então ao invés de copiar o que está no isin passado acima
vai ser copiado tudo menos o que está no isin, sendo denotado pelo simbolo "~"  '''
df_pvalue_filtrado = df[~df["educacao"].isin(ex_educacao)].copy()
df_pvalue_filtrado = df_pvalue_filtrado[~df_pvalue_filtrado["tipo_renda"].isin(ex_tipo_renda)].copy()
df_pvalue_filtrado = df_pvalue_filtrado[~df_pvalue_filtrado["estado_civil"].isin(ex_estado_civil)].copy()

In [14]:
# Usado para comparar pra ver se o que está nos códigos acima deu certo
df_pvalue_filtrado['estado_civil'].unique()

array(['Solteiro', 'Casado', 'Viúvo', 'Separado'], dtype=object)

In [15]:
# Usado para comparar pra ver se o que está nos códigos acima deu certo
df['estado_civil'].unique()

array(['Solteiro', 'Casado', 'Viúvo', 'União', 'Separado'], dtype=object)

In [16]:
# Removendo a variável tipo_residencia, que é a menos significante
y, X = patsy.dmatrices("np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel, Treatment(1)) + \
            qtd_filhos + C(tipo_renda) + C(educacao, Treatment('Secundário')) + \
            C(estado_civil) + idade + tempo_emprego + \
            qt_pessoas_residencia", df_pvalue_filtrado)

# Criando regressão com base no dmatrices sem o tipo_residencia
reg_log = sm.OLS(y, X).fit()
reg_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.355
Model:                            OLS   Adj. R-squared:                  0.355
Method:                 Least Squares   F-statistic:                     519.2
Date:                Mon, 14 Aug 2023   Prob (F-statistic):               0.00
Time:                        10:04:28   Log-Likelihood:                -14322.
No. Observations:               13211   AIC:                         2.867e+04
Df Residuals:                   13196   BIC:                         2.879e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
=============================================================================================================================
                                                                coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------------------------
Intercept                                                     6.6464      0.222     29.952      0.000       6.211       7.081
C(sexo)[T.M]                                                  0.8135      0.015     55.147      0.000       0.785       0.842
C(posse_de_veiculo)[T.True]                                   0.0380      0.014      2.729      0.006       0.011       0.065
C(posse_de_imovel, Treatment(1))[T.False]                    -0.0877      0.014     -6.473      0.000      -0.114      -0.061
C(tipo_renda)[T.Empresário]                                   0.1610      0.016     10.133      0.000       0.130       0.192
C(tipo_renda)[T.Pensionista]                                 -0.1882      0.022     -8.586      0.000      -0.231      -0.145
C(tipo_renda)[T.Servidor público]                             0.0552      0.023      2.387      0.017       0.010       0.101
C(educacao, Treatment('Secundário'))[T.Superior completo]     0.1159      0.013      8.795      0.000       0.090       0.142
C(estado_civil)[T.Separado]                                   0.2822      0.112      2.513      0.012       0.062       0.502
C(estado_civil)[T.Solteiro]                                   0.2493      0.110      2.262      0.024       0.033       0.465
C(estado_civil)[T.Viúvo]                                      0.3065      0.113      2.706      0.007       0.084       0.529
qtd_filhos                                                   -0.2633      0.110     -2.403      0.016      -0.478      -0.049
idade                                                         0.0050      0.001      6.227      0.000       0.003       0.007
tempo_emprego                                                 0.0621      0.001     57.353      0.000       0.060       0.064
qt_pessoas_residencia                                         0.2802      0.109      2.563      0.010       0.066       0.495
==============================================================================
Omnibus:                        0.748   Durbin-Watson:                   2.023
Prob(Omnibus):                  0.688   Jarque-Bera (JB):                0.716
Skew:                           0.008   Prob(JB):                        0.699
Kurtosis:                       3.033   Cond. No.                     2.46e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.46e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [17]:
x = df.shape[0]
x2= df_pvalue_filtrado.shape[0]

dados_removidos = x-x2
dados_removidos

1789

#### Os valores de R² e o R ajustado melhoraram quase que 1% e os valores de P estão apenas com variáveis significantes de acordo com o valor passado de menor que 5% e parece ter ficado melhor, apesar de que me incomoda muito a perca de dados e de acordo com o código visto acima foi perdido 1789 linhas nesse processo